userops_ext.go

```diff
@@ -0,0 +1,219 @@
+// by the wallet before the UserOperation is solved and it is a Read-Only field.
+type UserOperationExt struct {
+	OriginalHashValue string `json:"original_hash_value" mapstructure:"original_hash_value" validate:"required"`
+	ProcessingStatus  string `json:"processing_status" mapstructure:"processing_status" validate:"required"`
+}
+
+const IntentEndToken = "<intent-end>"
+
+type userOperationError string
+
+func (e userOperationError) Error() string {
+	return string(e)
+}
+
+// Define error constants
+const (
+	ErrNoIntentFound     userOperationError = "no Intent found"
+	ErrIntentInvalidJSON userOperationError = "invalid Intent JSON"
+	ErrNoSeparator       userOperationError = "separator token not found"
+	ErrNoCalldata        userOperationError = "no CallData found"
+)
+
+// extractIntentJSON attempts to extract the Intent JSON from the CallData field.
+// It returns the JSON as a string and a boolean indicating whether a valid JSON was found.
+func (u *UserOperation) extractIntentJSON() (string, bool) {
+	parts := strings.Split(string(u.CallData), IntentEndToken)
+	if len(parts) >= 1 {
+		var intent Intent
+		if err := json.Unmarshal([]byte(parts[0]), &intent); err == nil {
+			return parts[0], true
+		}
+	}
+	return "", false
+}
+
+// HasIntent checks if the CallData field contains a valid Intent JSON that
+// decodes successfully into an Intent struct.
+func (u *UserOperation) HasIntent() bool {
+	_, hasIntent := u.extractIntentJSON()
+	return hasIntent
+}
+
+// GetIntentJSON returns the Intent JSON from the CallData field, if present.
+func (u *UserOperation) GetIntentJSON() (string, error) {
+	intentJSON, hasIntent := u.extractIntentJSON()
+	if !hasIntent {
+		return "", ErrNoIntentFound
+	}
+	return intentJSON, nil
+}
+
+// GetIntent takes the Intent JSON from the CallData field, decodes it into
+// an Intent struct, and returns the struct.
+func (u *UserOperation) GetIntent() (*Intent, error) {
+	intentJSON, hasIntent := u.extractIntentJSON()
+	if !hasIntent {
+		return nil, ErrNoIntentFound
+	}
+
+	var intent Intent
+	if err := json.Unmarshal([]byte(intentJSON), &intent); err != nil {
+		return nil, ErrIntentInvalidJSON
+	}
+	return &intent, nil
+}
+
+// HasCallData returns true if the CallData field starts with "0x"
+// either directly or after the intent and separator token.
+func (u *UserOperation) HasCallData() bool {
+	parts := strings.Split(string(u.CallData), IntentEndToken)
+
+	// Check for CallData after the separator token
+	const hexPrefix = "0x"
+	if len(parts) >= 2 {
+		return strings.HasPrefix(parts[1], hexPrefix)
+	}
+
+	// Check for CallData directly if there's no separator token
+	return strings.HasPrefix(parts[0], hexPrefix)
+}
+
+// GetCallData extracts and returns the Ethereum transaction CallData from the CallData field.
+// It returns an error if the CallData value does not exist or does not start with "0x".
+func (u *UserOperation) GetCallData() ([]byte, error) {
```

Comment:
you're extracting call data from the call data.
I think it is misleading to name it calldata because calldata is already in use.
calldata = intent + execution.

Reply:
@EliasiOfir, I'll address this in another PR so we can test the merging release process now. 
Note this is the only comment in the PR.